### PR TITLE
Add network flushing before setting up network

### PR DESCRIPTION
**Pull request by sergio-correia**
This should allow for using the nbde_client role with machines that use
static IP configurations, as network flushing should undo the network
setup done at the initramfs, allowing the system to use its regular
configuration.

Approach based on the answers posted here:
https://unix.stackexchange.com/questions/506331/networkmanager-doesnt-change-ip-address-when-dracut-cmdline-provided-static-ip/541108

### CODE REVIEW

File: files/nbde_client-network-flush [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+
+for iface in $(ip link | sed -rn 's@^[0-9]+:\s+([[:alnum:]]+):.*@\1@p'); do
```
Comment (richm) [inline files/nbde_client-network-flush:3]: @tyll is there a better way to do this?

File: tasks/main-clevis.yml [hunk 2]
```diff
@@ -56,4 +68,26 @@
...
+    - name: Deploy network flushing script
+      copy:
+        src: "{{ role_path }}/files/nbde_client-network-flush"
+        dest: /usr/bin/nbde_client-network-flush
+        mode: '0755'
+
+    - name: Deploy network flushing service
+      copy:
+        src: "{{ role_path }}/files/nbde_client-network-flush.service"
+        dest: /etc/systemd/system/nbde_client-network-flush.service
```
Comment (richm) [inline tasks/main-clevis.yml:82]: I think using `mode: '0644'` will get rid of the warning

File: files/nbde_client-network-flush (hunk 1 above)
Comment (thom311) [reply files/nbde_client-network-flush:3]: JSON output of iproute2 would be nicer. But on some systems, that's not available (in particular, CentOS7).

From sysfs might be good. Note there are some special files there like `bonding_masters`. Guess you could do something like `(cd /sys/class/net/; for f in */; do echo ${f%/}; done)` to get rid of that.

but I have no strong preference here.

Sergio, did you strip interfaces with a parent, like `vethf6a43f04@if9`, on purpose?

Comment (sergio-correia) [reply files/nbde_client-network-flush:3]: > Sergio, did you strip interfaces with a parent, like `vethf6a43f04@if9`, on purpose?

It was probably accidental. I initially was using `nmcli -t device`, which was simpler to use the output with `cut`, but I think I had issues with CentOS 7, then I switched to this other approach. I like your suggestion (and tyll's) of using `/sys/class/net/`, so I will change to use it instead. Thanks!

File: tasks/main-clevis.yml (hunk 2 above)
Comment (sergio-correia) [reply tasks/main-clevis.yml:82]: Done.

**Comment by richm** [comment]
[citest bad]

**Comment by richm** [comment]
@sergio-correia how do we test this?  Can we have an automated test for this?

**Comment by richm** [comment]
[citest bad]

**Comment by sergio-correia** [comment]
> @sergio-correia how do we test this? Can we have an automated test for this?

One way to to test this would be to configure a system set up with a LUKS-encrypted device and a static IP address. We can then use the ansible role and once the machine does the automated unlocked, after a reboot, we can inspect that it still has the same static IP that it was configured before.

it's not impossible to have an automated test for this, but it's not straightforward either. We could easily provision a VM with the required setup with a kickstart file, for instance, but then we need a way to make sure the VM boots properly the first time, without user interaction.

**Comment by richm** [comment]
> > @sergio-correia how do we test this? Can we have an automated test for this?
> 
> One way to to test this would be to configure a system set up with a LUKS-encrypted device and a static IP address. We can then use the ansible role and once the machine does the automated unlocked, after a reboot, we can inspect that it still has the same static IP that it was configured before.

Do we need a separate nbde server machine to test this?  If so, then we can't have an automated test for this, at least with our current test framework.  Then assuming we can't have an automated test for this, what is the QE procedure to verify this change?

> 
> it's not impossible to have an automated test for this, but it's not straightforward either. We could easily provision a VM with the required setup with a kickstart file, for instance, but then we need a way to make sure the VM boots properly the first time, without user interaction.

**Comment by sergio-correia** [comment]
> Do we need a separate nbde server machine to test this? If so, then we can't have an automated test for this, at least with our current test framework. Then assuming we can't have an automated test for this, what is the QE procedure to verify this change?
> 

Yes, the nbde server machine cannot be the same one as the client, as the client will attempt to communicate with it while booting. If using virtualization, we could run the server on the host, whlie the guest would run the client.

As for the QE procedure to verify this, I am not 100% sure, but I suspect it would be a manual test doing what I suggested in the previous message: setting up a machine with LUKS and static IP and using the nbde_client role + a reboot to verify things worked as expected.

**Comment by richm** [comment]
lgtm - but first you will need to rebase on top of the latest master code - this changes the way the role is symlinked under tests/roles/linux-system-roles.nbde_client - you will need to add symlinks there for `files` and `templates`

File: tasks/main-clevis.yml [hunk 3]
```diff
@@ -4,6 +4,18 @@
...
+  service:
+    name: clevis-luks-askpass.path
+    enabled: yes
+
+- name: Generate nbde_client dracut config
+  template:
+    src: nbde_client.conf
+    dest: /etc/dracut.conf.d/nbde_client.conf
+    owner: root
+    mode: u=rw
```
Comment (sarroutbi) [inline tasks/main-clevis.yml:17]: Are write permissions required for this file? I guess 0444 permission should be enough

Comment (sergio-correia) [reply tasks/main-clevis.yml]: Done.

**Comment by richm** [comment]
> @richm: I have rebased it and added the symlinks.

Thanks!  I'm monitoring the test runs - if they look good I'll merge

**Comment by richm** [comment]
[citest]

**Pull request merged**